gen/map.go

```diff
@@ -190,6 +190,101 @@ func (m *mapGenerator) Reader(g Generator, spec *compile.MapSpec) (string, error
 	return name, wrapGenerateError(spec.ThriftName(), err)
 }
 
+// Equals generates a function to compare maps of the given type
+//
+// 	func $name(lhs, rhs $mapType) bool {
+// 		...
+// 	}
+//
+// And returns its name.
+func (m *mapGenerator) Equals(g Generator, spec *compile.MapSpec) (string, error) {
+	name := "_" + valueName(spec) + "_Equals"
+
+	err := g.EnsureDeclared(
+		`
+			<$mapType := typeReference .Spec>
+
+			<$lhs := newVar "lhs">
+			<$rhs := newVar "rhs">
+			func <.Name>(<$lhs>, <$rhs> <$mapType>) bool {
+				if len(<$lhs>) != len(<$rhs>) {
+					return false
+				}
+
+				<$lk := newVar "lk">
+				<$lv := newVar "lv">
+				<$rv := newVar "rv">
+				<$ok := newVar "ok">
+				<if isHashable .Spec.KeySpec>
+					for <$lk>, <$lv> := range <$lhs> {
+						<$rv>, <$ok> := <$rhs>[<$lk>]
+						if !<$ok> {
+							return false
```

Comment:
Given how significantly the bodies differ in the hashable versus unhashble case, consider doing this check outside the template and simply calling a different function.

```
if !isHashable(spec) {
  return m.EqualsUnhashble(g, spec)
}

err := g.TextFromTemplate(
  ...
)
```